### PR TITLE
ISPN-12402 + ISPN-12184 Multiple single-port endpoints and explicit/implicit fixes

### PR DESCRIPTION
**Pull request by tristantarrant**
This PR introduces the ability to have multiple single-port endpoints. It also allows disabling admin features (CLI/Console) per-endpoint. Additionally it fixes an issue where it was not really respecting explicit authentication mechanisms.

https://issues.redhat.com/browse/ISPN-12184
https://issues.redhat.com/browse/ISPN-12402

### CODE REVIEW

**Review by pruivo — CHANGES_REQUESTED** [review]
minor suggestions.

**Comment by pruivo** [comment]
```
17:06:52,033 INFO  (ForkJoinPool.commonPool-worker-3) [org.infinispan.SERVER] ISPN080018: Protocol HotRod (internal)
17:06:52,115 INFO  (main) [org.infinispan.SERVER] ISPN080018: Protocol REST (internal)
17:06:52,285 INFO  (main) [org.infinispan.SERVER] ISPN080004: Protocol SINGLE_PORT listening on 127.0.0.1:11222
17:06:52,285 INFO  (main) [org.infinispan.SERVER] ISPN080034: Server 'pedro-desktop-17255' listening on http://127.0.0.1:11222
17:06:52,302 INFO  (ForkJoinPool.commonPool-worker-3) [org.infinispan.SERVER] ISPN080018: Protocol HotRod (internal)
17:06:52,335 INFO  (main) [org.infinispan.SERVER] ISPN080018: Protocol REST (internal)
17:06:52,664 INFO  (main) [org.infinispan.SERVER] ISPN080004: Protocol SINGLE_PORT listening on 192.168.2.100:13333
17:06:52,664 INFO  (main) [org.infinispan.SERVER] ISPN080034: Server 'pedro-desktop-17255' listening on http://192.168.2.100:13333
17:06:52,664 INFO  (main) [org.infinispan.SERVER] ISPN080001: Infinispan Server 12.0.0-SNAPSHOT started in 5329ms
```

* I think there is a verb missing in  `ISPN080018` message
* `ISPN080004: Protocol SINGLE_PORT ` IMO, it would be nice to add the `socket-binding` name to the log message, like
`Protocol SINGLE_PORT (<socket-binding>) listening on 127.0.0.1:11222`

File: server/runtime/src/main/java/org/infinispan/server/configuration/endpoint/EndpointConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.infinispan.commons.configuration.attributes.AttributeDefinition;
+import org.infinispan.commons.configuration.attributes.AttributeSet;
+import org.infinispan.commons.configuration.elements.DefaultElementDefinition;
+import org.infinispan.commons.configuration.elements.ElementDefinition;
+import org.infinispan.server.configuration.Element;
+import org.infinispan.server.core.configuration.ProtocolServerConfiguration;
+import org.infinispan.server.router.configuration.SinglePortRouterConfiguration;
+
+/**
+ * @since 10.0
```
Comment (pruivo) [inline server/runtime/src/main/java/org/infinispan/server/configuration/endpoint/EndpointConfiguration.java:16]: nitpick:
```suggestion
 * @since 12.0
```

**Review by pruivo — CHANGES_REQUESTED** [review]
The JSON writer is broken and `JsonSerializationTest` is failing. 

It is generating the following json
```
"endpoints": {
      "endpoints": {
        "socket-binding": "default",
        "security-realm": "default",
        "hotrod-connector": {
```
 IMO, it should be an array of endpoints since you can have multiple endpoints now. 
```
"endpoints": [
      {
        "socket-binding": "default",
        "security-realm": "default",
        "hotrod-connector": {
```

I guess (not tested!) the easy fix would be `elements.add(endpoints);` in `ServerConfiguration` change to `elements.addAll(endpoints.subElements());`

File: server/runtime/src/main/java/org/infinispan/server/Server.java [hunk 2]
```diff
@@ -348,43 +355,48 @@ public void setExitHandler(ExitHandler exitHandler) {
...
+                        routes.put(new Route<>(routeSource, new HotRodServerRouteDestination(protocolServer.getName(), (HotRodServer) protocolServer)), 0);
+                        extensions.apply((HotRodServer) protocolServer);
+                     } else if (protocolServer instanceof RestServer) {
+                        routes.put(new Route<>(routeSource, new RestServerRouteDestination(protocolServer.getName(), (RestServer) protocolServer)), 0);
+                     }
+                     log.protocolStarted(protocolServer.getName());
                   }
-                  log.protocolStarted(protocolServer.getName());
+               } catch (Throwable t) {
+                  System.err.println(t.getMessage());
```
Comment (pruivo) [inline server/runtime/src/main/java/org/infinispan/server/Server.java:384]: I guess this requires a proper log message :)

File: server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java [hunk 3]
```diff
@@ -1102,10 +1102,10 @@ private void parseDataSourceConnectionPool(XMLExtendedStreamReader reader, DataS
    }
 
    private void parseEndpoints(XMLExtendedStreamReader reader, ConfigurationBuilderHolder holder, ServerConfigurationBuilder builder) throws XMLStreamException {
-      EndpointsConfigurationBuilder endpoints = builder.endpoints();
+
```
Comment (pruivo) [inline server/runtime/src/main/java/org/infinispan/server/configuration/ServerConfigurationParser.java:1105]: nitpick: remove empty line

File: server/tests/src/test/resources/configuration/endpoints/default-auth-implicit.xml [hunk 4]
```diff
@@ -2,6 +2,6 @@
            xsi:schemaLocation="urn:infinispan:server:12.0 https://infinispan.org/schemas/infinispan-server-12.0.xsd"
            xmlns="urn:infinispan:server:12.0"
            socket-binding="default" security-realm="default">
-   <hotrod-connector name="hotrod" />
-   <rest-connector name="rest" />
+   <hotrod-connector name="hotrod"/>
```
Comment (pruivo) [inline server/tests/src/test/resources/configuration/endpoints/default-auth-implicit.xml:5]: nit: you could remove the connector name for hotrod & rest

**Review by pruivo — CHANGES_REQUESTED** [review]
minor comment.

**Comment by pruivo** [comment]
@tristantarrant there are test failures, probably related:
* RestServerResource (NPE)
* CliIT

**Comment by pruivo** [comment]
offtopic: while testing "admin" and "metric-auth", it looks like the `disconnect` is broken. `connection = null` is missing in `org.infinispan.cli.impl.ContextImpl#disconnect()`. I guess it can be fixed here :) it is just one line.

```
pedro-desktop-53750@cluster//containers/default]> disconnect 
[pedro-desktop-53750@cluster//containers/default]> disconnect 
[pedro-desktop-53750@cluster//containers/default]> disconnect 
[pedro-desktop-53750@cluster//containers/default]> disconnect -h
Usage: disconnect [<options>]
Disconnects from a remote server

Options:
  -h, --help

[pedro-desktop-53750@cluster//containers/default]> disconnect 
[pedro-desktop-53750@cluster//containers/default]> ls
Could not connect to server: java.io.InterruptedIOException: executor rejected
[pedro-desktop-53750@cluster//containers/default]> ls
Could not connect to server: java.io.InterruptedIOException: executor rejected
[pedro-desktop-53750@cluster//containers/default]> cd 
The name of the subsystem/item is required for this command.
[pedro-desktop-53750@cluster//containers/default]> ls
Could not connect to server: java.io.InterruptedIOException: executor rejected
[pedro-desktop-53750@cluster//containers/default]> cd /
[pedro-desktop-53750@cluster/]> ls
containers
cluster
server
[pedro-desktop-53750@cluster/]> cd containers
[pedro-desktop-53750@cluster//containers]> ls
default
[pedro-desktop-53750@cluster//containers]> cd default
Could not connect to server: java.io.InterruptedIOException: executor rejected
```

**Comment by tristantarrant** [comment]
@pruivo I've addressed all issues I believe

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
integrated! thanks @tristantarrant !

**Pull request merged**